File: week2/mergeSort.py
def merge(A,B):
    (m,n) = (len(A),len(B))
    (C,i,j,k) = ([],0,0,0)
    while (k<m+n):
        if i==m:        # if A is empty
            C.extend(B[j:])     #add rest of B to c
            k = k + (n-j)       # change the index of k to 
        elif j==n:
            C.extend(A[i:])
            k = k + (m - i)
        elif A[i] < B[j]:
            C.append(A[i])
            i = i + 1
            k = k + 1
        else:
            C.append(B[j])
            j = j + 1
            k = k + 1
    return C

def mergeSort(L):
    n = len(L)
    if n <= 1:
        return L
    left = mergeSort(L[:n//2])
    right = mergeSort(L[n//2:])
    
    merged_and_sorted = merge(left,right)
    return merged_and_sorted

def main():
    arr = [1,-66,3,654,33,4,2,62,754,51,0,9]
    print(mergeSort(arr))

main()

File: week2/javascript/mergeSort.js
function merge(A, B) {
    let [m, n] = [A.length, B.length];
    let [C, i, j, k] = [[], 0, 0, 0];

    while (k < m + n) {
        if (i == m) { // A is empty
            C.push(...B.slice(j)); // push remaining of items from B to C
            k = k + (n - j); 
        } else if (j == n) {
            C.push(...A.slice(i));
            k = k + (m - i);
        } else if (A[i] < B[j]) {
            C.push(A[i]);
            i++;
            k++;
        } else {
            C.push(B[j]);
            j++;
            k++;
        }
    }

    return C;
}


function mergeSort(arr){
    let n = arr.length;
    if(n<=1) return arr;

    let mid = Math.floor(n/2);

    let left = mergeSort(arr.slice(0,mid));
    let right = mergeSort(arr.slice(mid));

    let merged = merge(left,right);

    return merged;
}



(function main(){
    let unsorted_arr = [1,0,0,0,-66,3,654,33,4,2,62,754,51,0,9];
    console.log(mergeSort(unsorted_arr));
}())